crates/relayer/src/chain/ckb4ibc/message/conn.rs

```diff
@@ -0,0 +1,297 @@
+    new_ibc_connection_cell.next_connection_number += 1;
+
+    let envelope = Envelope {
+        msg_type: MsgType::MsgConnectionOpenTry,
+        content: rlp::encode(&CkbMsgConnectionOpenTry {
+            proof: convert_proof(msg.proofs)?,
+        })
+        .to_vec(),
+    };
+
+    let envelope_encoded = get_encoded_object(envelope);
+    let old_connection_encoded = get_encoded_object(old_ibc_connection_cell);
+    let new_connection_encoded = get_encoded_object(new_ibc_connection_cell);
+
+    let data_capacity = Capacity::zero()
+        .safe_add(32u32)
+        .map_err(|_| Error::capacity())?;
+
+    let packed_tx = CoreTransactionView::new_advanced_builder()
+        .input(converter.get_ibc_connections_input())
+        .output(
+            CellOutput::new_builder()
+                .lock(
+                    Script::new_builder()
+                        .code_hash(converter.get_config().connection_code_hash.pack())
+                        .args(converter.get_config().connection_args.pack())
+                        .build(),
+                )
+                .build_exact_capacity(data_capacity.clone())
+                .map_err(|_| Error::capacity())?,
+        )
+        .output_data(new_connection_encoded.data)
+        .witness(
+            WitnessArgs::new_builder()
+                .input_type(old_connection_encoded.witness)
+                .output_type(new_connection_encoded.witness)
+                .build()
+                .as_bytes()
+                .pack(),
+        )
+        .witness(
+            WitnessArgs::new_builder()
+                .output_type(envelope_encoded.witness)
+                .build()
+                .as_bytes()
+                .pack(),
+        )
+        .build();
+    Ok(packed_tx.into())
+}
+
+pub fn convert_conn_open_ack_to_tx<C: MsgToTxConverter>(
+    msg: MsgConnectionOpenAck,
+    converter: &C,
+) -> Result<TransactionView, Error> {
+    let old_ibc_connection_cell = converter.get_ibc_connections();
+    let mut new_ibc_connection_cell = old_ibc_connection_cell.clone();
+
+    let idx = msg
+        .connection_id
+        .as_str()
+        .parse::<usize>()
+        .map_err(|_| Error::ckb_conn_id_invalid(msg.connection_id.to_string()))?;
+    let mut connection_end = new_ibc_connection_cell.connections.get_mut(idx).unwrap();
+    connection_end.state = State::Open;
+
+    let envelope = Envelope {
+        msg_type: MsgType::MsgConnectionOpenAck,
+        content: rlp::encode(&CkbMsgConnectionOpenAck {
+            conn_id_on_a: idx,
+            proof_conn_end_on_b: convert_proof(msg.proofs)?,
+        })
+        .to_vec(),
+    };
+    let envelope_encoded = get_encoded_object(envelope);
+    let old_connection_encoded = get_encoded_object(old_ibc_connection_cell);
+    let new_connection_encoded = get_encoded_object(new_ibc_connection_cell);
+
+    let data_capacity = Capacity::zero()
+        .safe_add(32u32)
+        .map_err(|_| Error::capacity())?;
```

Comment:
`data_capacity` should be calculated by the bytes length of output data, you can use `Capacity::bytes(...).unwrap()` to do so